### DISCUSSION TITLE
Bot randomly stops responding to messages then starts again after another random amount of time

### DISCUSSION THREAD

**Discussion by JackAshwell11** [Rapptz/discord.py#6142] [Q&A] [answered]
So I've created a bot which takes quotes from pre-existing text, uploads gifs, etc. But I'm trying to code in a trivia portion where the bot asks the user a question and the user has to respond with the correct answer. The code so far:

	import asyncio  
	import discord  
	import random  
	import AO3  
	import pandas as pd  
	  
	# Initialise discord variables  
	token = "TOKEN"  
	client = discord.Client()  
	  
	# Function to setup the dataframe  
	def dataframeSetup():  
		# Create the dataframe  
		df = pd.read_csv("https://docs.google.com/spreadsheets/d/16QtBJEtvV5a5DheR78x5AsoVA5b2DpXD1mq-x3lCFiA/export?format=csv", names=["NaN", "Title", "Author", "Ship(s)", "Type", "Series", "Status", "Smut", "No of words", "No of chapters", "Link"])  
		# Remove first two lines  
		df = df.iloc[2:]  
		# Remove the first column  
		df.drop("NaN", axis=1, inplace=True)  
		# Create variable to store the index of the first empty row  
		firstEmptyRow = 0  
		# Iterate over every row  
		for index, row in df.iterrows():  
			# Test if every cell is empty  
			if row.isnull().all():  
				# Set firstEmptyRow to the index (it is minus 2 because the index of the dataframe starts at 2)  
				firstEmptyRow = index - 2  
				break  
		# Return the final dataframe  
		return df.iloc[0:firstEmptyRow]  
	  
	# Initialise the dataframe  
	df = dataframeSetup()  
	# Initialise trivia variables  
	df1 = pd.read_csv("Trivia.txt", delimiter="/", names=["Question", "Option 1", "Option 2", "Option 3", "Option 4", "Correct Option"])  
	triviaDone = []  
	  
	# Function to make random quotes  
	def quoteMaker(df):  
		# Grab a random fic  
		randomFic = df.iloc[random.randint(2, len(df))]  
		# Create AO3 session  
		ao3Session = AO3.Session("username", "password")  
		# Create work object  
		work = AO3.Work(AO3.utils.workid_from_url(randomFic["Link"]), ao3Session)  
		# Get chapter ammount  
		chapterAmmount = work.chapters  
		# Get chapter text for a random chapter  
		randomChapter = random.randint(1, chapterAmmount)  
		randomChapterText = work.get_chapter_text(randomChapter)  
		# Convert the chapter text into a list  
		textList = list(filter(None, randomChapterText.split("\n")))  
		# Create list to assign the authors to  
		result = []  
		if len(work.authors) > 1:  
			# More than one author  
			for author in work.authors:  
				result.append(author.username)  
		else:  
			# Only one author  
			result.append(work.authors[0].username)  
		# Return random string  
		return textList[random.randint(0, len(textList)-1)], work, randomChapter, ', '.join(result)  
	  
	# Function to create trivia  
	def triviaMaker(triviaDone):  
		# Test if all trivia questions have been done  
		if len(triviaDone) == len(df1):  
			# They've all been done, so clear the list and start again  
			triviaDone.clear()  
		# Generate a random index and use that to get a random trivia question  
		randomIndex = random.randint(0, len(df1))-1  
		randomTrivia = df1.iloc[randomIndex]  
		# Test if the selected trivia question has been done before  
		while randomIndex in triviaDone:  
			# Trivia has already been done recently so try another one  
			randomTrivia = df.iloc[random.randint(0, len(df1))]  
		# Add the selected trivia question's index to the list  
		triviaDone.append(randomIndex)  
		# Return the formatted string as well as the correct index to allow for validation  
		return f'''{randomTrivia["Question"]}:  
		1. {randomTrivia["Option 1"]}  
		2. {randomTrivia["Option 2"]}  
		3. {randomTrivia["Option 3"]}  
		4. {randomTrivia["Option 4"]}''', randomTrivia, randomTrivia["Correct Option"]  
	  
	# Detect new message  
	@client.event  
	async def on_message(message):  
		if message.content == "!quote":  
			# Grab quote and related attributes  
			quote, work, randomChapter, authors, = quoteMaker(df)  
			# Print quote and attributes  
			await message.channel.send(quote)  
			await message.channel.send(f"-{work.title} chapter {randomChapter} by {authors}. Link {work.url}")  
		elif message.content == "!trivia":  
			# Display trivia question  
			triviaString, randomTrivia, correctIndex = triviaMaker(triviaDone)  
			# Store channel id  
			channel = message.channel  
			await message.channel.send(triviaString)  
			# Function to check if an answer is correct  
			def check(message):  
				# Check if answer is correct  
				if "!answer" in message.content:  
					return message.content == f"!answer {randomTrivia[correctIndex]}" or message.content == f"!answer {correctIndex}"  
			# Try and except statement to catch timeout error  
			try:  
				# Wait for user response  
				await client.wait_for("message", check=check, timeout=15)  
				# User response is correct  
				await channel.send("Correct answer")  
			except asyncio.TimeoutError:  
				# User has run out of time  
				await channel.send(f"Times up, correct answer was option {correctIndex}. {randomTrivia[correctIndex]}")  
		elif message.content == "!cruzie":  
			# User has types !cruzie so do secret  
			await message.channel.send("https://giphy.com/gifs/midland-l4FsJgbbeKQC8MGBy")  
		elif message.content == "!murica":  
			# User has typed !murica so play murica gif  
			await message.channel.send("https://tenor.com/view/merica-gif-9091003")  
	  
	# Run when discord bot has started  
	@client.event  
	async def on_ready():  
		# Get channel ID  
		channel = client.get_channel(channel)  
		# Send message to user signalling that the bot is ready  
		await channel.send("Running")  
	  
	# Start discord bot  
	client.run(token)

However, sometimes after like 5 messages, it'll stop then start up again after a couple minutes but other times it'll be after 15 message and start up again in like an hours time.

Anyone know what the problem is? If so, how could I detect that it's happening so I can send a message to the channel.

P.S. Sorry for my bad formatting, I'm not use to github markdown

**Comment by Nemika-Haj** [comment, marked as answer]
Could you use proper formatting? It's hard to understand if any of this is right, indent-wise.

**Reply by JackAshwell11** [reply]
Just updated it. Sorry

**Reply by Nemika-Haj** [reply]
@Aspect1103 have you checked if your bot is being rate-limited or something?

**Reply by JackAshwell11** [reply]
How do I check that?

**Reply by Nemika-Haj** [reply]
You can try to run a command while the bot is not responding to the trivia question. If it does not work, then the bot is being rate-limited.

**Reply by JackAshwell11** [reply]
Is there a way to detect before that happens so the users are just stuck there typing commands in with nothing happening?

**Reply by Nemika-Haj** [reply]
Well, you see, the bot retrieves data from every single message sent. I suggest making a proper command handler. This way it will probably limit the requests and the bot will run smoothly. For the time, check if rate-limiting is the issue. If it is, I can guide you through fixing it.

**Reply by JackAshwell11** [reply]
If you can, that'll be great. Thanks.

And how would I make a better command handler?

**Reply by Nemika-Haj** [reply]
Basically, add this below `import discord`
```py
import discord
from discord.ext import commands #<--- This
```

Then you can replace 
```py
client = discord.Client()
```
with
```py
client = commands.Bot(command_prefix="the-prefix-you-desire", case_insensitive=True)
# This will basically create a Bot client. Its prefix will be the value of command_prefix. case_insensitive is so the bot ignores command capitalization.
```

Then you can take each `if` statement you have made inside the `on_message` event and create a new command. To create a new command, use this format;
```py
@client.command()
async def quote(ctx):
    quote, work, randomChapter, authors, = quoteMaker(df)
    await ctx.send(quote)
    await ctx.send(f"-{work.title} chapter {randomChapter} by {authors}. Link {work.url}")
```
This is the `!quote` command you have. You can follow the same logic for each one. The `ctx` parameter is required for all commands! You can find more information [here](https://discordpy.readthedocs.io/en/latest/ext/commands/api.html?highlight=context#discord.ext.commands.Context).

**Reply by JackAshwell11** [reply]
How would I get an !answer command to sync with the !trivia command then since !trivia has a client.wait_for?

**Reply by Nemika-Haj** [reply]
You do not need to, keep your check function as it is and it will work. Just remember to change `message` to `ctx.message`

**Reply by JackAshwell11** [reply]
Yeah, I tried that:

	# Register !trivia command
	@client.command()
	async def trivia(ctx):
		# Display trivia question
		triviaString, randomTrivia, correctIndex = triviaMaker(triviaDone)
		# Store channel id
		channel = ctx.channel
		await ctx.channel.send(triviaString)
		# Function to check if an answer is correct
		def check(message):
			# Check if answer is correct
			if "!answer" in ctx.message.content:
				return ctx.message.content == f"!answer {randomTrivia[correctIndex]}" or ctx.message.content == f"!answer {correctIndex}"
		# Try and except statement to catch timeout error
		try:
			# Wait for user response
			await client.wait_for("message", check=check, timeout=15)
			# User response is correct
			await ctx.channel.send("Correct answer")
		except asyncio.TimeoutError:
			# User has run out of time
			await ctx.channel.send(f"Times up, correct answer was option {correctIndex}. {randomTrivia[correctIndex]}")

But I keep getting the error:

	Ignoring exception in command None:
	discord.ext.commands.errors.CommandNotFound: Command "answer" is not found

**Reply by Nemika-Haj** [reply]
Does the command work though? If it does, that's good. Now you can create an error handler in order to hide the error, or just an empty command (a command that just returns nothing once executed).

**Reply by JackAshwell11** [reply]
It does work, but I don't know where I would put the error handler? Or is an empty command better?

**Reply by Nemika-Haj** [reply]
Both work. An error handler is recommended for future use.

You can create an `on_command_error` function and check if the `error` is an instance of `commands.CommandNotFound`. If it is, simply `return`. Otherwise, `raise error`.

An empty command would be simpler. Just create a `@client.command()` method, add `async def answer(ctx): return`, and you're done.
```py
'''
 Command Handler
What's going on here is that we're creating a `client.event` method.
`on_command_error` takes two arguments (as it can be found here;
https://discordpy.readthedocs.io/en/latest/ext/commands/api.html?highlight=on_command_error#discord.ext.commands.Bot.on_command_error
Then, we check if `error` is an instance of `commands.CommandNotFound`, which is our raised error when calling `!answer`.
If it is an instance, it will simply do nothing.
Otherwise, it will raise the error.
'''
@client.event
async def on_command_error(ctx, error):
    if isinstance(error, commands.CommandNotFound): return
    else: raise error

'''
  Empty Command
There you simply create a command to do... nothing. xd
'''
@bot.command()
async def answer(ctx): return #basically nothing.

**Reply by JackAshwell11** [reply]
Great, thanks a ton. You've been really helpful

**Reply by Nemika-Haj** [reply]
You're welcome! Remember to mark this as answered if your issue was resolved.
